File: opensso/products/federation/library/source/com/sun/identity/saml2/protocol/RequestedAuthnContext.java
/**
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS HEADER.
 *
 * Copyright (c) 2006 Sun Microsystems Inc. All Rights Reserved
 *
 * The contents of this file are subject to the terms
 * of the Common Development and Distribution License
 * (the License). You may not use this file except in
 * compliance with the License.
 *
 * You can obtain a copy of the License at
 * https://opensso.dev.java.net/public/CDDLv1.0.html or
 * opensso/legal/CDDLv1.0.txt
 * See the License for the specific language governing
 * permission and limitations under the License.
 *
 * When distributing Covered Code, include this CDDL
 * Header Notice in each file and include the License file
 * at opensso/legal/CDDLv1.0.txt.
 * If applicable, add the following below the CDDL Header,
 * with the fields enclosed by brackets [] replaced by
 * your own identifying information:
 * "Portions Copyrighted [year] [name of copyright owner]"
 *
 * $Id: RequestedAuthnContext.java,v 1.2 2008/06/25 05:47:57 qcheng Exp $
 *
 */


package com.sun.identity.saml2.protocol;

import java.util.List;

import org.w3c.dom.Element;

import com.sun.identity.saml2.common.SAML2Exception;

/**
 * Java content class for RequestedAuthnContext element declaration.
 * <p>The following schema fragment specifies the expected 	
 * content contained within this java content object. 	
 * <p>
 * <pre>
 * &lt;element name="RequestedAuthnContext" type="{urn:oasis:names:tc:SAML:2.0:protocol}RequestedAuthnContextType"/>
 * </pre>
 * 
 * @supported.all.api
 */
public interface RequestedAuthnContext {

    /**
     * Returns list of authentication context class references. References
     * in the list is String object.
     *
     * @return list of authentication context class references.
     */
    List getAuthnContextClassRef();
    
    /**
     * Sets authentication context class references.
     *
     * @param references List of authentication context class references where
     *        references in the list is String object.
     * @throws SAML2Exception if this object is immutable.
     */
    void setAuthnContextClassRef(List references)
        throws SAML2Exception;

    /**
     * Returns list of authentication context declaration class references. 
     * References in the list is String object.
     *
     * @return list of authentication context declaration class references.
     */
    List getAuthnContextDeclRef();

    /**
     * Sets authentication context declaration class references.
     *
     * @param references List of authentication context declaration class 
     *        references where references in the list is String object.
     * @throws SAML2Exception if this object is immutable.
     */
    void setAuthnContextDeclRef(List references) throws SAML2Exception;

    /**
     * Returns the value of <code>Comparison</code> property.
     * 
     * @return the value of <code>Comparison</code> property.
     */
    String getComparison();

    /**
     * Sets the value of the <code>Comparison</code> property.
     * 
     * @param value the value of the <code>Comparison</code> property.
     * @throws SAML2Exception if <code>Object</code> is immutable.
     */
    void setComparison(String value) throws SAML2Exception;

    /**
     * Returns a String representation of this Object.
     * 
     * @return a String representation of this Object.
     * @throws SAML2Exception if it failed to pasrse the object.
     */
    String toXMLString() throws SAML2Exception;
	
    /**
     * Returns a String representation of this Object.
     *
     * @param includeNSPrefix true to prepend the namespace qualifier
     *        to the Element.
     * @param declareNS true to declare the namespace within the Element.
     * @return the String representation of this Object.
     * @throws SAML2Exception if it failed to pasrse the object.
     */ 
    String toXMLString(boolean includeNSPrefix, boolean declareNS)
           throws SAML2Exception;

    /** 
    * Makes this object immutable by making this object unmodifiable.
    */
    void makeImmutable() ;
   
    /** 
    * Returns true if mutable, false otherwise.
    *
    * @return true if mutable, false otherwise.
    */
    boolean isMutable();
}

File: openam-federation/openam-federation-library/src/main/csharpsource/Fedlet/Fedlet/source/Saml2/RequestedAuthnContext.cs
/*
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS HEADER.
 * 
 * Copyright (c) 2009 Sun Microsystems Inc. All Rights Reserved
 * 
 * The contents of this file are subject to the terms
 * of the Common Development and Distribution License
 * (the License). You may not use this file except in
 * compliance with the License.
 * 
 * You can obtain a copy of the License at
 * https://opensso.dev.java.net/public/CDDLv1.0.html or
 * opensso/legal/CDDLv1.0.txt
 * See the License for the specific language governing
 * permission and limitations under the License.
 * 
 * When distributing Covered Code, include this CDDL
 * Header Notice in each file and include the License file
 * at opensso/legal/CDDLv1.0.txt.
 * If applicable, add the following below the CDDL Header,
 * with the fields enclosed by brackets [] replaced by
 * your own identifying information:
 * "Portions Copyrighted [year] [name of copyright owner]"
 * 
 * $Id: RequestedAuthnContext.cs,v 1.1 2009/06/11 18:37:58 ggennaro Exp $
 */
/*
 * Portions Copyrighted 2013 ForgeRock Inc.
 */

using System;
using System.Collections;
using System.Text;
using Sun.Identity.Properties;
using Sun.Identity.Saml2.Exceptions;

namespace Sun.Identity.Saml2
{
    /// <summary>
    /// Class representing the RequestedAuthnContext.
    /// </summary>
    public class RequestedAuthnContext
    {
        #region Members
        #endregion

        #region Constructors
        /// <summary>
        /// Initializes a new instance of the RequestedAuthnContext class.
        /// </summary>
        public RequestedAuthnContext()
        {
            this.Comparison = null;
            this.AuthnContextClassRef = new ArrayList();
            this.AuthnContextDeclRef = new ArrayList();
        }
        #endregion

        #region Properties
        /// <summary>
        /// Gets or sets the Comparison.
        /// </summary>
        public string Comparison { get; set; }

        /// <summary>
        /// Gets the AuthnContextClassRef.
        /// </summary>
        public ArrayList AuthnContextClassRef { get; private set; }

        /// <summary>
        /// Gets the AuthnContextDeclRef.
        /// </summary>
        public ArrayList AuthnContextDeclRef { get; private set; }

        #endregion

        #region Methods

        /// <summary>
        /// Generates the XML string of the RequestedAuthnContext using
        /// the Comparison, AuthnContextClassRef, and AuthnContextDeclRef
        /// information.
        /// </summary>
        /// <returns>Returns the RequestedAuthnContext XML as a string.</returns>
        public string GenerateXmlString()
        {
            if (this.AuthnContextClassRef.Count == 0 && this.AuthnContextDeclRef.Count == 0)
            {
                throw new Saml2Exception(Resources.RequestedAuthnContextClassRefOrDeclRefNotDefined);
            }

            if (String.IsNullOrEmpty(this.Comparison))
            {
                this.Comparison = "exact";
            }
            else if (!this.ValidComparison())
            {
                throw new Saml2Exception(Resources.RequestedAuthnContextInvalidComparison);
            }

            StringBuilder rawXml = new StringBuilder();

            rawXml.Append("<samlp:RequestedAuthnContext Comparison=\"");
            rawXml.Append(this.Comparison);
            rawXml.Append("\">");

            if (this.AuthnContextClassRef != null)
            {
                foreach (string value in this.AuthnContextClassRef)
                {
                    rawXml.Append("<saml:AuthnContextClassRef>");
                    rawXml.Append(value);
                    rawXml.Append("</saml:AuthnContextClassRef>");
                }
            }

            if (this.AuthnContextDeclRef != null)
            {
                foreach (string value in this.AuthnContextDeclRef)
                {
                    rawXml.Append("<saml:AuthnContextDeclRef>");
                    rawXml.Append(value);
                    rawXml.Append("</saml:AuthnContextDeclRef>");
                }
            }

            rawXml.Append("</samlp:RequestedAuthnContext>");

            return rawXml.ToString();
        }

        /// <summary>
        /// Sets the AuthnContextClassRef list.
        /// </summary>
        /// <param name="list">The list to become the AuthnContextClassRef.</param>
        public void SetAuthnContextClassRef(ArrayList list)
        {
            if (list == null)
            {
                this.AuthnContextClassRef = new ArrayList();
            }
            else
            {
                this.AuthnContextClassRef = list;
            }
        }

        /// <summary>
        /// Sets the SetAuthnContextDeclRef list.
        /// </summary>
        /// <param name="list">The list to become the SetAuthnContextDeclRef.</param>
        public void SetAuthnContextDeclRef(ArrayList list)
        {
            if (list == null)
            {
                this.AuthnContextDeclRef = new ArrayList();
            }
            else
            {
                this.AuthnContextDeclRef = list;
            }
        }

        /// <summary>
        /// Checks to see if the Comparison property is set to a valid
        /// value based on the SAML specification.
        /// </summary>
        /// <returns>True if it is a valid comparison, false otherwise.</returns>
        private bool ValidComparison()
        {
            return this.Comparison == "exact"
                || this.Comparison == "maximum"
                || this.Comparison == "minimum"
                || this.Comparison == "better";
        }

        #endregion
    }
}